### PR TITLE
Update Cloud Function with Upload Processor 

### PR DESCRIPTION
**Pull request by vivbak**
Workflow is as follows
1. gVCFs and CRAMS are moved.
2. Following a successful move, analysis objects are created.
3. Once analysis objects are successfully created, sequencing metadata is updated

### CODE REVIEW

**Comment by vivbak** [comment]
Updated Workflow: 
`main.py` defines an endpoint, that when hit, formats a file object. This file object, and the associated qc_metadata is added to the sample metadata db. 
`process_backlog.py` exists to process previous batches.

File: update_sample_status/main.py [hunk 1]
```diff
@@ -1,63 +1,120 @@
...
-    return ('', 204)
+
+    if not project or not external_id or not status or not batch or not qc_metadata:
+        raise web.HTTPBadRequest(
+            reason=f'Invalid request. Ensure that the project, sample, status, batch and metadata are provided.'
+        )
+
+    # Determine the internal ID
+    external_id_dict = {'external_ids': [external_id]}
+    internal_id_map = sapi.get_sample_id_map_by_external(project, external_id_dict)
```
Comment (illusional) [inline update_sample_status/main.py:80]: This recently changed, so you don't need to wrap it in a dictionary anymore.

```suggestion
    internal_id_map = sapi.get_sample_id_map_by_external(project, [external_id])
```

File: update_sample_status/main.py [hunk 2]
```diff
@@ -1,63 +1,120 @@
...
+
+    reads = []
+    for key in FILES_HANDLED:
+        reads = create_file_object(external_id, key, project, batch)
+        # TODO: Added the QC Metadata Here. Should this be moved to sample metadata?
+        metadata = {'reads': reads, 'qc_metrics': qc_metadata}
+        # Update the sequencing metadata.
+        update_sequence_meta(project, internal_id, status, metadata)
+
+    return web.Response(text=f'{external_id} lodged.')
```
Comment (illusional) [inline update_sample_status/main.py:91]: It looks like you set `reads` to an empty array, and then set on every iteration, but I think it'd be better to not set everything against reads (as initially suggested), and instead against the different types:

- `reads` for (`.cram`)
- `gvcf` for (`.gvcf`)

If you add `'meta-field'` to the FILES_HANDLED dict, your code could look something like:

```suggestion
    # TODO: Added the QC Metadata Here. Should this be moved to sample metadata?
    meta = {'qc_metrics': qc_metadata}
    for key in FILES_HANDLED:
        reads = create_file_object(external_id, key, project, batch)
        
        field_name = FILES_HANDLED[key]['meta-field']
        meta[field_name] = reads
        meta[field_name + '_type'] = key
        
    # Update the sequencing metadata.
    update_sequence_meta(project, internal_id, status, metadata)

    return web.Response(text=f'{external_id} lodged.')
```

File: update_sample_status/main.py [hunk 3]
```diff
@@ -1,63 +1,120 @@
...
+from sample_metadata.models import SequencingStatus
 
 
-secret_manager = secretmanager.SecretManagerServiceClient()
+routes = web.RouteTableDef()
 
+FILES_HANDLED = {
+    'gvcf': [{'primary': '.g.vcf.gz', 'secondary': ['.g.vcf.gz.tbi', '.g.vcf.gz.md5']}],
+    'cram': [{'primary': '.cram', 'secondary': ['.cram.crai', '.cram.md5']}],
+}
```
Comment (illusional) [inline update_sample_status/main.py:23]: To address future comment.

```suggestion
FILES_HANDLED = {
    'gvcf': [{'primary': '.g.vcf.gz', 'secondary': ['.g.vcf.gz.tbi', '.g.vcf.gz.md5'], 'meta-field': 'gvcf'}],
    'cram': [{'primary': '.cram', 'secondary': ['.cram.crai', '.cram.md5'], 'meta-field': 'reads'}],
}
```

**Review by illusional — CHANGES_REQUESTED** [review]
This is looking pretty good!

**Pull request closed without merging**